File: app/src/main/java/se/gustavkarlsson/skylight/android/cache/DualAuroraReportSingletonCache.java
package se.gustavkarlsson.skylight.android.cache;

import android.content.Context;

import com.vincentbrison.openlibraries.android.dualcache.Builder;
import com.vincentbrison.openlibraries.android.dualcache.DualCache;

import se.gustavkarlsson.skylight.android.BuildConfig;
import se.gustavkarlsson.skylight.android.models.AuroraReport;

public class DualAuroraReportSingletonCache implements SingletonCache<AuroraReport> {
	private static final String KEY = "singleton"; // Must match [a-z0-9_-]{1,64}

	private final DualCache<AuroraReport> dualCache;

	public DualAuroraReportSingletonCache(Context context, String cacheName) {
		Builder<AuroraReport> builder = new Builder<AuroraReport>(cacheName, BuildConfig.VERSION_CODE)
				.useReferenceInRam(Integer.MAX_VALUE, object -> 1)
				.useSerializerInDisk(Integer.MAX_VALUE, false, new GsonCacheSerializer<>(AuroraReport.class), context);
		if (BuildConfig.DEBUG) {
			builder.enableLog();
		}
		this.dualCache = builder.build();
	}

	@Override
	public AuroraReport getValue() {
		return dualCache.get(KEY);
	}

	@Override
	public void setValue(AuroraReport report) {
		dualCache.put(KEY, report);
	}

}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/cache/DualAuroraReportSingletonCache.kt
package se.gustavkarlsson.skylight.android.cache

import android.content.Context
import com.vincentbrison.openlibraries.android.dualcache.Builder
import com.vincentbrison.openlibraries.android.dualcache.DualCache
import se.gustavkarlsson.skylight.android.BuildConfig
import se.gustavkarlsson.skylight.android.models.AuroraReport

private const val KEY = "singleton" // Must match [a-z0-9_-]{1,64}

class DualAuroraReportSingletonCache(context: Context, cacheId: String) : SingletonCache<AuroraReport> {
    private val dualCache: DualCache<AuroraReport>

    init {
        val builder = Builder<AuroraReport>(cacheId, BuildConfig.VERSION_CODE)
                .useReferenceInRam(Integer.MAX_VALUE) { 1 }
                .useSerializerInDisk(Integer.MAX_VALUE, false, GsonCacheSerializer(AuroraReport::class.java), context)
        if (BuildConfig.DEBUG) {
            builder.enableLog()
        }
        this.dualCache = builder.build()
    }

    override var value: AuroraReport?
        get() = dualCache.get(KEY)
        set(report) = dualCache.put(KEY, report)

}
